### PR TITLE
samples: mesh: single timer to handle all transitions,  upgrade the implementation of Generic Move Message handlers & bugs fix

### PR DESCRIPTION
**Pull request by vikrant8052**
First commit upgrade the implementation to reduce complexity in
algorithm. Previously there was different timers for lightness
as well as temperature transitions. But at time only single
transition was remain activated & hence there is no need
of multiple timer for each entity.

If Node Lightness is in downward transition & received Generic
OnOff message to set state equal to 1 then Node transaction get
stopped in between. Ideally it should reach to default lightness
value if it is non-zero. Second commit has solved this bug.

After receiving Generic Level Move set/set_unack message, generic
level state should move towards positive or negative extreme end
(which is depend upon sign of received delta value) till it not get interrupted.
4th commit has introduced this feature.

Signed-off-by: Vikrant More <vikrant8051@gmail.com>

### CODE REVIEW

File: samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/device_composition.c [hunk 1]
```diff
@@ -71,31 +71,21 @@ BT_MESH_MODEL_PUB_DEFINE(gen_level_cli_pub_s0, NULL, 2 + 7);
 /* Definitions of models publication context (End) */
 
 /* Definitions of models user data (Start) */
-struct generic_onoff_state gen_onoff_srv_root_user_data = {
-	.transition = &lightness_transition,
-};
+struct generic_onoff_state gen_onoff_srv_root_user_data;
```
Comment (jhedberg) [inline samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/device_composition.c]: Sinne these are global variables (non-static), the above change means that it is not anymore initialized here. Where are you then initializing it?

File: samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/device_composition.c [hunk 2]
```diff
@@ -2677,4 +2699,3 @@ const struct bt_mesh_comp comp = {
 	.elem = elements,
 	.elem_count = ARRAY_SIZE(elements),
 };
-
```
Comment (jhedberg) [inline samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/device_composition.c]: How is this change related to the commit?

File: samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/publisher.c [hunk 3]
```diff
@@ -241,4 +241,3 @@ void publish(struct k_work *work)
 		printk("bt_mesh_model_publish: err: %d\n", err);
 	}
 }
-
```
Comment (jhedberg) [inline samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/publisher.c]: How is this change related to the commit?

File: samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/state_binding.c [hunk 4]
```diff
@@ -305,4 +305,3 @@ void calculate_temp_target_values(u8_t type)
 			light_ctl_srv_user_data.delta_uv;
 	}
 }
-
```
Comment (jhedberg) [inline samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/state_binding.c]: How is this change related to the commit?

**Review by jhedberg — CHANGES_REQUESTED** [review]

File: samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/device_composition.c (hunk 1 above)
Comment (vikrant8051) [reply samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/device_composition.c]: No need to initialize them here.

But some variables under those structures are initialized under **light_default_var_init()**
defined in **main.c** file.

They will re-initialized under light_default_status_init() which will get called after reading
variable stored under "/ps" settings branch.

Comment (vikrant8051) [reply samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/device_composition.c]: Other variable will be initialized as zero. And they will get automatically initialized
further once the state_binding() function get called in light_default_status_init(void).

Kindly refer main.c for details.

File: samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/device_composition.c (hunk 2 above)
Comment (vikrant8051) [reply samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/device_composition.c]: I think it was get automatically removed because of using atom editor. 
I've added line as it was. So done !!

File: samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/publisher.c (hunk 3 above)
Comment (vikrant8051) [reply samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/publisher.c]: done !!

File: samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/state_binding.c (hunk 4 above)
Comment (vikrant8051) [reply samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/state_binding.c]: done !!

**Comment by vikrant8051** [comment]
@jhedberg Please merge the PR if everything is perfect.

**Comment by vikrant8051** [comment]
wait

**Comment by vikrant8051** [comment]
After fine tuning & testing now everything is perfect.

**Comment by vikrant8051** [comment]
wait

File: samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/device_composition.c (hunk 1 above)
Comment (vikrant8051) [reply samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/device_composition.c]: Done !!

**Comment by vikrant8051** [comment]
https://3pl46c46ctx02p7rzdsvsg21-wpengine.netdna-ssl.com/wp-content/uploads/2019/04/1903_Mesh-Models-Overview_FINAL.pdf

After reading this document, I am going to upgrade implementation of Generic Level Message handlers.
I am going to push latest changes soon.

**Comment by jhedberg** [comment]
> After reading this document, I am going to upgrade implementation of Generic Level Message handlers. I am going to push latest changes soon.

@vikrant8051 that's fine as long as it doesn't make this PR any bigger. The main issue which is causing slowness in getting your PRs merged is that they are just too big for a thorough review. So if you're going to do more extensive changes please consider splitting them up into multiple PRs.

**Comment by vikrant8051** [comment]
@jhedberg 
I have added 4th commit to correct implementation of only generic move set/set_unack message handlers.

`Generic move set - initiates changing the generic level state in either a positive or negative     direction and at a given speed. The speed with which the transition takes place is calculated from        a       delta level field in the message and a value known as the transition time. Transition time must     either appear in the generic move set message itself, where it is an optional parameter, or be     available in a state called the generic default transition time, which belongs to the generic default      transition time model which may or may not be present. If transition time is not available from     either of these two sources, the operation will not be executed and generic level will not      be changed.`


`The move transition may continue indefinitely. It will stop if a move set message with the delta         level field set to zero is received. When generic level reaches its upper or lower limit, during a     move transition, the implementation may decide to either terminate the transition at that point or     take some other action, such as wrapping around and continuing`

As per old implementation, there was no significant difference between gen. delta & gen. move
message handlers. But now I have correctly interpreted what exactly Mesh Model specification is suggesting.
  
Other things are already tested & Okay from my side. Please review it.

**Comment by vikrant8051** [comment]
> `
The main issue which is causing slowness in getting your PRs merged is that they are just too big for a thorough review.`

@jhedberg I have edited commit 1 so that it will not aggressively impact old implementation & hope it
will now help in fast review.

**Comment by vikrant8051** [comment]
recheck

**Comment by jhedberg** [comment]
@mfiumara FYI, it seems @vikrant8051 added a patch here which does essentially the same thing as your PR #15933

File: samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/main.c [hunk 5]
```diff
@@ -110,8 +109,8 @@ static void light_default_status_init(void)
 
 	default_tt = gen_def_trans_time_srv_user_data.tt;
 
-	target_lightness = lightness;
-	target_temperature = temperature;
+	init_lightness_target_values();
+	init_temp_target_values();
```
Comment (jhedberg) [inline samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/main.c]: Why do you do the initialisation at runtime instead of build time? It seems all this code is called from main() before loading settings, i.e. you always know the state up-front, i.e. there's probably a lot of redundant code in these init functions (in the branches that will never be taken)

File: samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/state_binding.c [hunk 6]
```diff
@@ -224,6 +224,26 @@ void state_binding(u8_t light, u8_t temp)
...
+	}
+
+	gen_level_srv_root_user_data.target_level = target_lightness + INT16_MIN;
+
+	light_lightness_srv_user_data.target_actual = target_lightness;
+
+	light_lightness_srv_user_data.target_linear =
+		actual_to_linear(target_lightness);
+
+	light_ctl_srv_user_data.target_lightness = target_lightness;
```
Comment (jhedberg) [inline samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/state_binding.c]: As I mentioned earlier, all of the initialisation done by this function should be possible to do at build time rather than runtime. Only way this would make a little bit sense is if you called this code *after* loading settings, but that's not what the current code is doing (it's calling settings_load() from bt_ready, which is called after these init functions)

**Review by jhedberg — CHANGES_REQUESTED** [review]

File: samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/main.c (hunk 5 above)
Comment (vikrant8051) [reply samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/main.c]: No. Please refer main.c, I have called light_default_status_init() after loading settings.

Comment (jhedberg) [reply samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/main.c]: I don't see that. I see that you call bt_enable() asyncrhously, passing it the bt_ready callback. You then proceed with calling light_default_status_init() right after that, i.e. without any guarantee that bt_ready (which calls settings_load) will have been called.

Comment (vikrant8051) [reply samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/main.c]: As of now, let it be. I will take care of it in next PR (Since it is going on from very long time).
If you have quick suggestions then please suggest it so that I can quickly repair the things.

Comment (jhedberg) [reply samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/main.c]: You could call bt_enable synchronously by passing NULL to it, and then explicitly call bt_ready from main() once bt_enabled has returned success.

Comment (vikrant8051) [reply samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/main.c]: Thanks for the suggestion. I've added 6th commit as per it.

File: samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/main.c [hunk 7]
```diff
@@ -197,9 +197,15 @@ void main(void)
...
 
 	/* Initialize the Bluetooth Subsystem */
-	err = bt_enable(bt_ready);
+	err = bt_enable(NULL);
 	if (err) {
 		printk("Bluetooth init failed (err %d)\n", err);
+		return;
+	} else {
+		bt_ready(err);
+		if (err) {
```
Comment (jhedberg) [inline samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/main.c]: err is 0 in this branch, so this test doesn't make sense

**Review by jhedberg — CHANGES_REQUESTED** [review]
One comment in the code. Please also fix the commit message. This is not about the error handling,  rather about synchronising the init procedure to initialise the mesh state only after the settings have been loaded.

File: samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/main.c (hunk 7 above)
Comment (jhedberg) [reply samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/main.c]: you can also remove the err parameter from bt_ready since you're never passing anything else than 0

Comment (vikrant8051) [reply samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/main.c]: Done !!

File: samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/ble_mesh.c [hunk 8]
```diff
@@ -51,7 +51,7 @@ static const struct bt_mesh_prov prov = {
 	.reset = prov_reset,
 };
 
-void bt_ready(int err)
+void bt_ready(void)
 {
 	struct bt_le_oob oob;
```
Comment (jhedberg) [inline samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/ble_mesh.c]: CI will probably fail since you're still using err inside this function, i.e. you'll need to add a local err variable

Comment (vikrant8051) [reply samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/ble_mesh.c]: yes !!

Comment (vikrant8051) [reply samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/ble_mesh.c]: have just push new request

**Comment by Vudentz** [comment]
Looks like there is a lot (168) of unrelated patches included in this PR, it seems some rebase is needed.
